odoo/osv/expression.py

```diff
@@ -882,33 +708,38 @@ def push_result(leaf):
 
             elif len(path) > 1 and field.store and field.type == 'many2one' and field.auto_join:
                 # res_partner.state_id = res_partner__state_id.id
-                leaf.add_join_context(comodel, path[0], 'id', path[0])
-                push(create_substitution_leaf(leaf, (path[1], operator, right), comodel))
+                coalias, _ = self.query.add_join(
+                    (alias, comodel._table, path[0], 'id', path[0]),
+                    implicit=False, outer=True,
+                )
+                push((path[1], operator, right), comodel, coalias)
 
             elif len(path) > 1 and field.store and field.type == 'one2many' and field.auto_join:
                 # res_partner.id = res_partner__bank_ids.partner_id
-                leaf.add_join_context(comodel, 'id', field.inverse_name, path[0])
+                coalias, _ = self.query.add_join(
+                    (alias, comodel._table, 'id', field.inverse_name, path[0]),
+                    implicit=False, outer=True,
+                )
+                push((path[1], operator, right), comodel, coalias)
+
```

Comment:
Might be a good idea to merge https://github.com/odoo/odoo/pull/48494 in there, as auto_join seem pretty much broken for o2ms (and likely m2ms).

Reply:
I prefer to make this proposal fix one problem.

I think your proposal is acceptable for the case of one2many fields with `auto_join`, but not in the general case, where it bypasses every possible override of `search` or `_search` (there are 15 such overrides in our modules).  The (possibly not documented) goal of `auto_join` is *performance*, at the price of bypassing access rights and search overrides.

I have another proposal for the general case: make `_search` and `_name_search` return the query to perform, so that this query can be injected as a subquery into the search.  This one will be the subject of a future PR.